src/tokens/Assets.tsx

```diff
@@ -0,0 +1,425 @@
+              : -scrollPosition.value,
+        },
+      ],
+      shadowColor: interpolateColor(
+        scrollPosition.value,
+        [nonStickyHeaderHeight - 10, nonStickyHeaderHeight + 10],
+        ['transparent', 'rgba(48, 46, 37, 0.15)']
+      ),
+    }
+  }, [scrollPosition.value, nonStickyHeaderHeight])
+
+  const animatedFooterStyles = useAnimatedStyle(() => {
+    return {
+      transform: [
+        {
+          translateY: footerPosition.value,
+        },
+      ],
+    }
+  }, [footerPosition.value])
+
+  useScrollAwareHeader({
+    navigation,
+    title: t('totalAssets'),
+    subtitle:
+      !tokensAreStale && totalBalanceLocal.gte(0)
+        ? t('totalBalanceWithLocalCurrencySymbol', {
+            localCurrencySymbol,
+            totalBalance: totalBalanceLocal.toFormat(2),
+          })
+        : `${localCurrencySymbol} -`,
+    scrollPosition,
+    startFadeInPosition: nonStickyHeaderHeight - HEADER_OPACITY_ANIMATION_START_OFFSET,
+    animationDistance: HEADER_OPACITY_ANIMATION_DISTANCE,
+  })
+
+  const handleMeasureNonStickyHeaderHeight = (event: LayoutChangeEvent) => {
+    setNonStickyHeaderHeight(event.nativeEvent.layout.height)
+  }
+
+  const handleMeasureListHeaderHeight = (event: LayoutChangeEvent) => {
+    setListHeaderHeight(event.nativeEvent.layout.height)
+  }
+
+  const handleMeasureListFooterHeight = (event: LayoutChangeEvent) => {
+    setListFooterHeight(event.nativeEvent.layout.height)
+  }
+
+  const handleChangeActiveView = (_: string, index: number) => {
+    navigation.setParams({ activeTab: index })
+    ValoraAnalytics.track(
+      [
+        AssetsEvents.view_wallet_assets,
+        AssetsEvents.view_collectibles,
+        AssetsEvents.view_dapp_positions,
+      ][index]
+    )
+  }
+
+  const tokenItems = useMemo(() => tokens.sort(sortByUsdBalance), [tokens])
+  const positionSections = useMemo(() => {
+    const positionsByDapp = new Map<string, Position[]>()
+    positions.forEach((position) => {
+      if (positionsByDapp.has(position.appName)) {
+        positionsByDapp.get(position.appName)?.push(position)
+      } else {
+        positionsByDapp.set(position.appName, [position])
+      }
+    })
+
+    const sections: SectionListData<TokenBalance | Position, SectionData>[] = []
+    positionsByDapp.forEach((positions, appName) => {
+      sections.push({
+        data: positions,
+        appName,
+      })
+    })
+    return sections
+  }, [positions])
+
+  const renderSectionHeader = ({
+    section,
+  }: {
+    section: SectionListData<TokenBalance | Position, SectionData>
+  }) => {
+    if (section.appName) {
+      return (
+        <View style={styles.positionSectionHeaderContainer}>
+          <Text style={styles.positionSectionHeaderText}>
+            {section.appName.toLocaleUpperCase()}
+          </Text>
+        </View>
+      )
+    }
+    return null
+  }
+
+  const renderAssetItem = ({ item }: { item: TokenBalance | Position }) => {
```

Comment:
I understand this logic is largely copied, but this isn't the easiest thing to wrap my head around (having `item` being a union and then disambiguating with a type guard).

Reply:
yeah I didn't find this easy to read as well. The renderSectionHeader function also does something similar. The item being union type starts from this type here https://github.com/valora-inc/wallet/blob/0e5eb6cf6a7786570aba3dcc6cc861eefc6eb800/src/tokens/Assets.tsx#L63-L65

So even if we make two separate list instances with specific functions, the renderAssetItem function's `item` type will be of the union type. I attempted making the `AnimatedSectionList` a generic type but the changes were getting large, so I retained the existing code. One thing we could do is make two explicit `AnimatedSectionList` with specific types, but let me know if you have other ideas